ASP.NET Metrics Dashboard for SigNoz Integration

karanveersingh5623 opened the pull request:
[https://github.com/SigNoz/signoz/issues/5996](https://github.com/SigNoz/signoz/issues/5996)
/claim [https://github.com/SigNoz/signoz/issues/5996](https://github.com/SigNoz/signoz/issues/5996)
Add aspnet dashboard - oltp, v1

# ASP.NET Metrics Dashboard for SigNoz Integration

## Description:
This PR introduces an ASP.NET metrics dashboard with integrated OpenTelemetry (OTel) instrumentation to monitor application performance and telemetry data sent to SigNoz. Key features include:

### 1. Telemetry Data Collection
- Utilizes OpenTelemetry packages for ASP.NET, process, runtime, and core instrumentation to gather metrics and traces, which are then exported to SigNoz.

### 2. OTLP Configuration
- Configures OTLP (OpenTelemetry Protocol) for data ingestion using gRPC and HTTP receivers on ports `4317` and `4318`, allowing compatibility with various communication protocols.

### 3. Exporting Data to SigNoz
- Sets up OTLP exporter to forward collected telemetry data, including process-level metrics (CPU, memory) and application-specific metrics (latency, error rates), to SigNoz, with environment-specific configurations.

### 4. Dashboard Panels:
- **Application Performance**:
  - CPU and memory usage, request rate, and response times.
- **Request and Response Behavior**:
  - Tracks request latency, error rates, and active requests.

### 5. Environment and Resource Attributes:
- Identifies application environment (Staging/Production) for filtered monitoring in the SigNoz dashboard.

This integration enhances observability by automating metrics and trace collection and centralizing insights into the ASP.NET application's performance.

karanveersingh5623 committed:
Add aspnet dashboard - oltp, v1

makeavish commented:
Can you please add the `service.name` as well?
Also any specific reason to use manual instrumentation, zero code instrumentation should also work

karanveersingh5623 commented:
@makeavish 

For the service.name, I added a new environment variable for the service name. Users can filter the metrics based on the service name. I'm not sure if this update fulfills the requirement or not.

for zero-code instrumentation it didn't work as expected. It seems like the collected data was not accurate.
Zero-code instrumentation is fast and easy to set up, but in our use case, it didn't collect the metrics that I want to build a dashboard on. So, manual instrumentation is needed.

The attached image shows the list of metrics that we can use to monitor our application. The documentation of each library contains the manual setup for the app. So basically, we need to add more code to get the metrics that we want.
https://opentelemetry.io/docs/zero-code/net/instrumentations/#metrics-instrumentations

Let me know if I am missing something

makeavish commented:
In most of the panels service.name filter is missing in WHERE clause

Also in below panel y-axis unit might be wrong, I think it's bytes
<img width="1370" alt="Screenshot 2024-11-27 at 17 39 32" src="https://github.com/user-attachments/assets/142b9027-2da7-48f4-aceb-a586819f7dbf">

CPU Usage panel query seems to be incorrect.

Unit of Processors panel seems to be wrong
<img width="810" alt="Screenshot 2024-11-27 at 17 48 19" src="https://github.com/user-attachments/assets/a56fd0c8-61f2-4f16-abb4-2b7f31cec065">

Error rate query is incorrect
<img width="1380" alt="Screenshot 2024-11-27 at 17 57 25" src="https://github.com/user-attachments/assets/011a06ed-f6ec-445d-af31-4da287c83da9">

Request rate query is also incorrect
<img width="722" alt="Screenshot 2024-11-27 at 17 58 23" src="https://github.com/user-attachments/assets/7b907aea-581a-459d-b22d-e97ef0004696">

Suggest you to read this doc before fixing: https://signoz.io/docs/metrics-management/types-and-aggregation/?utm_source=product&utm_medium=query-builder#histogram--exponential-histogram

karanveersingh5623 committed:
asp.net V2

karanveersingh5623 commented:
@makeavish 

I tried to get the CPU usage but it seemed to not have the available metrics for it. So I switched to total CPU time for different states.

![image](https://github.com/user-attachments/assets/12a5c58b-6878-447a-89de-f729056cddc1)

For the Unit of Processors panel, I changed the aggregate operator to "Count". According to the docs, the Count operator will give us a number of something so it will fit well for displaying the number of processors

![image](https://github.com/user-attachments/assets/dc62fc17-fcf3-4227-92e3-2293e3ebd2fe)

For the error rate and request rate, the metric that I use ishttp_server_request_duration_countwhich is the only available metric to track the number of requests. I also switch to the "increase" operator, which will give us the total number of events. for example, I simulated the 404 error response code by sending a request 1000 times per second and it gave back a correct number.
![image](https://github.com/user-attachments/assets/97606fbd-7055-4438-985f-a7fa7d3c3ba0)
![image](https://github.com/user-attachments/assets/36d54dae-5c4f-4bf6-a9aa-7935c416bd1d)

Nageshbansal commented:
Hey @karanveersingh5623 ,
Thank you for your contributions to the dashboard bug bounty programme and for being part of the SigNoz community.

Sadly, we're winding down the bug bounty programme and closing all issues raised under it, including this one. The programme has been a wonderful way to collaborate with the community, and the dashboards that have come through it have made SigNoz materially better for everyone who uses it.

We thank you for the time and care that went into every contribution.
